### PR TITLE
porting calc_shoc_varorcovar

### PR DESCRIPTION
**Pull request by tcclevenger**
Implements the `calc_shoc_varorcovar` fortran subroutine in C++ and adds a bfb test to existing property test. 

I followed the structure of the `calc_shoc_vertflux`  subroutine (implemented in https://github.com/E3SM-Project/scream/pull/499) almost identically since the functions are so similar. Basically reusing the same bfb test.

### CODE REVIEW

File: components/scream/src/physics/shoc/shoc_calc_shoc_varorcovar.cpp
Comment (jeff-cohere) [inline components/scream/src/physics/shoc/shoc_calc_shoc_varorcovar.cpp:8]: Because these kinds of files tend to get copy-pasted a lot, we've been encouraging people to eliminate specific text ("for doing variance or covariance on Reals", in this case) and just keep comments generic.

File: components/scream/src/physics/shoc/shoc_calc_shoc_varorcovar_impl.hpp
Comment (ambrad) [inline components/scream/src/physics/shoc/shoc_calc_shoc_varorcovar_impl.hpp:40]: Small naming quibble: `(up_)grid1,2` sound like they are variables describing the grid, but in fact they are `invar1,2` values. So you might name them something like `invar1_s`, `invar1_sm1`, where `s` is the conventional index for the pack slot and `sm1` is for `s-1`.

Comment (tcclevenger) [reply components/scream/src/physics/shoc/shoc_calc_shoc_varorcovar_impl.hpp]: I agree with your suggestion. Names have been changed.

**Comment by ambrad** [comment]
I marked this WIP so the AT doesn't keep retesting. @tcclevenger, rm "[WIP]" from the title once you've fixed the non-BFB errors.

**Comment by tcclevenger** [comment]
Found the error: missing a `,value` for the declaration of the `tunefac` variable for the fortran code. Only showed up when calling c++ from fortran, so not in the BFB test. All tests passed on my workstation and weaver.

**Comment by ambrad** [comment]
@tcclevenger sometimes adding the AT:RETEST label is necessary to trigger the AT again. Not sure that was needed here, but I added the label in case.

**Comment by bartgol** [comment]
@tcclevenger , some AT context for reference:

The RETEST label forces the AT to bypass the rule that says "wait X hours after a fail before retesting". In SCREAM, we set X=4h. The reason is that a failure might be due to issues with the system, so waiting a bit can reduce risk of another failure that is unrelated to the actual code. On one hand, reducing failures that are not our fault help keeping the github issue log shorter; on the other hand, unless the git sha changes (i.e., new commits) the AT stops testing your PR after 3 consecutive failures (and you would need to add RETEST label), so removing those unrelated failures helps to keep testing going.

In this case, however, I _think_ the AT won't test the PR until it is approved. The reason is that the PR is not "trusted", so the AT refuses to test the PR until someone approves it (this is for security reasons, since the tests will run on our SNL machines). To be trusted, the PR need to refer to a branch in the main repo *or* the user must appear in a whitelisted team (our only whitelisted team is 'scream-team', as of now). The reason why main repo is always ok is because it means the user has write privileges on the repo, so it is inherently trusted.

This PR is from a fork, so it must mean Conrad has not been added to the scream-team yet (hence, he's not "trusted").

More info [here](https://acme-climate.atlassian.net/wiki/spaces/NGDNA/pages/1047397034/Pull+request+Auto-Tester+AT).

**Comment by bartgol** [comment]
Since you are in the scream-team, it's completely up to you.

**Comment by tcclevenger** [comment]
I've changed the fortran subroutine to use `bfb_square()` instead of `()**2`. Also, I changed the bfb test to make it more sensitive (so that it failed) and then made an order of operation change in the C++ implementation.

File: components/scream/src/physics/shoc/tests/shoc_varorcovar_tests.cpp
Comment (jgfouca) [inline components/scream/src/physics/shoc/tests/shoc_varorcovar_tests.cpp:277]: The 4th arg is undocumented here.

File: components/scream/src/physics/shoc/tests/shoc_varorcovar_tests.cpp
Comment (jgfouca) [inline components/scream/src/physics/shoc/tests/shoc_varorcovar_tests.cpp:286]: Use cxx11-style iteration.

File: components/scream/src/physics/shoc/tests/shoc_varorcovar_tests.cpp
Comment (jgfouca) [inline components/scream/src/physics/shoc/tests/shoc_varorcovar_tests.cpp:302]: Use cxx11-style iteration.

File: components/scream/src/physics/shoc/tests/shoc_varorcovar_tests.cpp
Comment (jgfouca) [inline components/scream/src/physics/shoc/tests/shoc_varorcovar_tests.cpp:308]: Use cxx11-style iteration.

**Review by jgfouca — CHANGES_REQUESTED** [review]
Looks good. Just a couple updates needed to the unit test in order to be conformant with the latest style.

File: components/scream/src/physics/shoc/tests/shoc_varorcovar_tests.cpp
Comment (tcclevenger) [reply components/scream/src/physics/shoc/tests/shoc_varorcovar_tests.cpp]: I consolidated all the outer-loops into 1. Let me know what you think.

**Comment by tcclevenger** [comment]
@bartgol Should the Autotester start running again now that the PR has an approval and the retest label has been applied?

**Comment by bartgol** [comment]
I think something is off with this PR. I think the AT is picking up wrong info on the last test, so it keeps skipping testing. Let me check with the AT devs.

**Comment by allevin** [comment]
@tcclevenger @bartgol I think you need to have @jgfouca address the "requested changes" in the inspection.  The AT is saying that the PR 

* "**THE LAST COMMIT TO THIS PULL REQUEST HAS BEEN REVIEWED, BUT NOT ACCEPTED OR REQUIRES CHANGES**"

This means that while the PR has been reviewed, it cannot move forward because someone did not accept the change -OR- that they requested changes...  

When all else fails... Always blame @jgfouca   :-)

File: components/scream/src/physics/shoc/tests/shoc_varorcovar_tests.cpp
Comment (jgfouca) [inline components/scream/src/physics/shoc/tests/shoc_varorcovar_tests.cpp:284]: This loop structure diverges from the standard BFB unit test pattern. It is slightly more efficient in terms of lines of code, but there's value in sticking to the expected pattern.

File: components/scream/src/physics/shoc/tests/shoc_varorcovar_tests.cpp
Comment (jgfouca) [inline components/scream/src/physics/shoc/tests/shoc_varorcovar_tests.cpp:292]: An explicit copy construction would be cleaner here.

Also, we should not get into the habit of doing the cxx Data copy after retrieving the f90 data. This will break if there is inout data.

**Review by jgfouca — CHANGES_REQUESTED** [review]

**Comment by tcclevenger** [comment]
@jgfouca I switched back to the old pattern I had for the bfb test, only now with c++11 style loops.  Basically following the new version of the vertflux bfb test.

**Review by jgfouca — APPROVED** [review]
Thanks!

**Pull request merged**
